### PR TITLE
enable manually redeploying without specifying versions

### PR DESCRIPTION
**Pull request by andreer**

### CODE REVIEW

File: controller-server/src/main/java/com/yahoo/vespa/hosted/controller/restapi/application/ApplicationApiHandler.java [hunk 1]
```diff
@@ -862,9 +862,30 @@ private HttpResponse deploy(String tenantName, String applicationName, String in
...
+        if(deployDirectly && !applicationPackage.isPresent() && !applicationVersion.isPresent() && !vespaVersion.isPresent()) {
+            // Redeploy the existing deployment with the same versions.
+            Optional<Deployment> deployment = controller.applications().get(applicationId)
+                    .map(Application::deployments)
+                    .flatMap(deployments -> Optional.ofNullable(deployments.get(zone)));
+
+            if(!deployment.isPresent())
+                throw new IllegalArgumentException("Can't redeploy application, no deployment currently exist");
+
+            applicationVersion = Optional.ofNullable(deployment.get().applicationVersion());
```
Comment (jonmv) [inline controller-server/src/main/java/com/yahoo/vespa/hosted/controller/restapi/application/ApplicationApiHandler.java]: `.applicationVersion()` never returns `null`. What you want to check for is whether it's `ApplicationVersion.unknown`.

File: controller-server/src/main/java/com/yahoo/vespa/hosted/controller/restapi/application/ApplicationApiHandler.java [hunk 2]
```diff
@@ -862,9 +862,30 @@ private HttpResponse deploy(String tenantName, String applicationName, String in
...
+            if(!applicationVersion.isPresent())
+                throw new IllegalArgumentException("Can't redeploy application, application version is unknown");
+
+            vespaVersion = Optional.ofNullable(deployment.get().version());
```
Comment (jonmv) [inline controller-server/src/main/java/com/yahoo/vespa/hosted/controller/restapi/application/ApplicationApiHandler.java]: Also never `null`.

**Review by jonmv — CHANGES_REQUESTED** [review]
One little thing to fix, but looks good otherwise.

Of course, it leaves even more mess — hopefully enough to make me clean it all up soon ┏( ﾟ益ﾟ)ゞ

**Review by jonmv — APPROVED** [review]
(╭ರ_•́)

**Comment by mpolden** [comment]
Merge?

**Comment by andreer** [comment]
Yes, thank you.

**Pull request merged**